pkg/notary/image_signer.go

```diff
@@ -0,0 +1,203 @@
+	if err != nil {
+		return err
+	}
+
+	remoteStore, err := s.makeRemoteStore(url, gun, caCertFilePath, keychain)
+	if err != nil {
+		return err
+	}
+
+	cryptoService, err := s.makeCryptoService(notarySecretDir)
+	if err != nil {
+		return err
+	}
+
+	for _, target := range targets {
+		repo, err := s.Factory.GetRepository(url, gun, remoteStore, cryptoService)
+		if err != nil {
+			return err
+		}
+
+		err = repo.PublishTarget(target)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *ImageSigner) makeGUNAndTargets(report lifecycle.ExportReport, keychain authn.Keychain) (data.GUN, []*client.Target, error) {
+	gun := data.GUN("")
+	var targets []*client.Target
+	for _, tag := range report.Image.Tags {
+		s.Logger.Printf("Signing tag '%s'\n", tag)
+		ref, err := name.ParseReference(tag, name.WeakValidation)
+		if err != nil {
+			return "", nil, err
+		}
+
+		s.Logger.Printf("Pulling image '%s'\n", ref.Context().Name()+"@"+report.Image.Digest)
+		image, _, err := s.Client.Fetch(keychain, ref.Context().Name()+"@"+report.Image.Digest)
+		if err != nil {
+			return "", nil, err
+		}
+
+		imageSize, err := image.Size()
+		if err != nil {
+			return "", nil, err
+		}
+
+		digestBytes, err := hex.DecodeString(strings.TrimPrefix(report.Image.Digest, "sha256:"))
+		if err != nil {
+			return "", nil, err
+		}
+
+		curGUN := data.GUN(ref.Context().Name())
+		if gun == "" {
+			gun = curGUN
+		} else if gun != curGUN {
+			return "", nil, errors.New("signing to multiple registries is not supported")
+		}
+
+		targets = append(targets, &client.Target{
+			Name: ref.Identifier(),
+			Hashes: map[string][]byte{
+				notary.SHA256: digestBytes,
+			},
+			Length: imageSize,
+		})
+	}
+
+	return gun, targets, nil
+}
+
+func (s *ImageSigner) makeRemoteStore(serverURL string, gun data.GUN, caCertFilePath string, keychain authn.Keychain) (storage.RemoteStore, error) {
+	transport := http.DefaultTransport
+
+	if caCertFilePath != "" {
+		s.Logger.Printf("Using CA certifiacte '%s'\n", caCertFilePath)
+
+		cert, err := ioutil.ReadFile(caCertFilePath)
+		if err != nil {
+			return nil, err
+		}
+
+		rootCAs := x509.NewCertPool()
+		if ok := rootCAs.AppendCertsFromPEM(cert); !ok {
+			return nil, errors.New("failed to load notary CA certificate")
+		}
+
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: rootCAs,
+			},
+			TLSHandshakeTimeout: 30 * time.Second,
+		}
+	}
+
+	return storage.NewNotaryServerStore(
+		serverURL,
+		gun,
+		&AuthenticatingRoundTripper{
+			Keychain:            keychain,
+			WrappedRoundTripper: transport,
+		},
+	)
+}
+
+func (s *ImageSigner) makeCryptoService(notarySecretDir string) (*cryptoservice.CryptoService, error) {
+	cryptoStore := storage.NewMemoryStore(nil)
+
+	fileInfos, err := ioutil.ReadDir(notarySecretDir)
+	if err != nil {
+		return nil, err
+	}
+
+	privateKeyFound := false
+	for _, info := range fileInfos {
+		if strings.HasSuffix(info.Name(), ".key") {
```

Comment:
Just look for any key in the configured secret?

Reply:
Right now, it is assumed that there is only one key.

Reply:
In the future, we could just parse each key file in the volume, as the keys are well formed and contain the GUN.